File: python/binarysearch.py
# 분할정복 방식의 이진 탐색 알고리즘
# 전제조건이 정렬이기 때문에 O(logN) 의 시간복잡도를 갖는다.
# python 에서는 bisect 라이브러리로 간단하게 쓸 수 있도록 제공한다

from bisect import bisect_left

# LeetCode 704 Binary Search
# 반복문 구현
def binary_search(self, nums, target):
    left, right = 0, len(nums) - 1

    while left <= right:
        mid = (left + right) // 2   # // 연산자를 사용해야 한다
        if nums[mid] == target:
            return mid
        elif target < nums[mid]:
            left = mid + 1
        else:
            rright = mid - 1

    return - 1 # 타겟이 배열에 없는 경우

# 재귀 구현
def search(self, nums, target):
    def bin_search(left, right):
        if left > right:
            return -1 # 타겟이 없는 경우
        
        mid = (left + right) // 2
        if nums[mid] == target:
            return mid
        elif nums[mid] < target:
            return bin_search(mid + 1, right)
        else:
            return bin_search(left, mid - 1)

    return bin_search(0, len(nums) - 1)


# bisect library 사용예
class Solution:
    def search(self, nums, target):
        index = bisect_left(nums, target)
        if index != len(nums) and nums[index] == target:
            return index
        return -1 # 타겟이 배열에 없는 경우


nums = [2, 6, 9, 15, 20, 50, 58, 67, 82]
nums.sort()
print("Sorted List : {nums}")

File: javalang/BinarySearch.java
package javalang;

// 절반씩 나눠계산하는 선형탐색의 한종류(완전탐색은 아니다)
// '정렬'을 해야한다는 전제 조건이 있다
// 시간복잡도는 O(logn) 이다
// BinarySearch 는 주로 문제해결을 위한 '수단'으로만 활용되기에 코드 템플렛을 외워두는 편이 낫다

// LeetCode 704 Binary Search
// 반복문 구현
public class BinarySearch {
    public int search(int[] nums, int target) {
        int left = 0, right = nums.length - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2;

            if (nums[mid] == target) {
                return mid;
            } else if (nums[mid] < target) {
                left = mid +1;
            } else {
                right = mid -1;
            }
        }
        return -1; // 타겟이 배열에 없는 경우
    }
}

// 재귀 구현
class BinarySearchRecursion {
    public int search(int[] nums, int target) {
        return binarySearch(nums, 0, nums.length -1, target);
    }

    private int binarySearch(int[] nums, int left, int right, int target) {
        if (left > right) {
            return -1; // 타겟이 배열에 없는 경우
        }
        
        int mid = left + (right - left) / 2;

        if (nums[mid] == target) {
            return mid;
        } else if (nums[mid] < target) {
            return binarySearch(nums, mid + 1, right, target);
        } else {
            return binarySearch(nums, left, mid - 1, target);
        }
    }
}